rootmap/sqlrootmap.go

```diff
@@ -19,38 +20,173 @@ SQLRootmap is a rootmap implementation backed by a sql.DB. The only database
 that has been used or tested is sqlite.
 */
 
+type sqlVersionStore struct {
+	reservationSize int
+	max             uint64
+	initialized     bool
+	mtx             sync.Mutex
+	c               uint64
+}
+
 type sqlRootmap struct {
 	db  *sql.DB
 	mtx *sync.Mutex
+	vs  *sqlVersionStore
+
+	tables map[table]bool
 }
 
-func (rm *sqlRootmap) initialize() error {
+func (rm *sqlRootmap) initialize(ctx context.Context) error {
+	rm.mtx.Lock()
+	defer rm.mtx.Unlock()
 	var maxApplied int64
 	err := rm.db.QueryRow("select max(version) from schema_migrations").Scan(&maxApplied)
 	if err == nil && maxApplied == 1 {
 		return nil
 	}
 	if _, err := rm.db.Exec(`
-	create table if not exists rootmap (
+	PRAGMA foreign_keys = ON;
+	create table if not exists tables (
+		id integer primary key,
 		database text not null,
-		producer_id text not null,
-		topic text not null,
+		producer text not null,
+		topic text not null
+	);
+
+	create unique index tables_producer_topic_unique on tables(database, producer, topic);
+
+	create table if not exists rootmap (
+		id integer primary key,
+		table_id integer not null references tables(id),
 		version bigint not null,
 		storage_prefix text not null,
 		node_id text not null,
-		timestamp text not null default current_timestamp,
-		primary key (producer_id, topic, version)
+		timestamp text not null
 	);
 
-	create table schema_migrations(
+	create index rootmap_table_id_idx on rootmap(table_id);
+
+	create table if not exists schema_migrations (
 		version bigint not null,
 		timestamp text not null default current_timestamp
 	);
 
 	insert into schema_migrations(version) values (1);
+
+	create table if not exists truncations (
+		table_id int not null references tables(id),
+		timestamp text not null default current_timestamp,
+		version bigint not null
+	);
+
+	create index truncations_table_id_timestamp_idx on truncations(table_id, timestamp);
+
+	create view latest_truncations as
+	select t1.*
+	from truncations t1
+	left join truncations t2
+	on t1.table_id = t2.table_id and t1.timestamp < t2.timestamp
+	where t2.table_id is null;
 	`); err != nil {
 		return fmt.Errorf("failed to migrate database: %w", err)
 	}
+	if err := rm.initializeVersionStore(ctx); err != nil {
+		return fmt.Errorf("failed to initialize version store: %w", err)
+	}
+	if err := rm.initializeTableCache(); err != nil {
+		return fmt.Errorf("failed to initialize table cache: %w", err)
+	}
+	return nil
+}
+
+func (rm *sqlRootmap) hasTable(database, producer, topic string) bool {
+	_, ok := rm.tables[table{database, producer, topic}]
```

Comment:
the access to tables here, and the write on 119, need to be guarded with a mutex.

I would suggest a separate, dedicated mutex from the one currently in use, and use a sync.RWMutex with just read locks here. The locks will be much shorter duration and primarily ready-only, relative to the ones used against sqlite.

Reply:
err, I guess strictly speaking it's the write on 220 that really requires synchronization. The one on 119 probably runs in serial since it's on startup.

IMO the cleanest way to handle this is to extract a "putTable" similar to your "hasTable" here, have those functions handle the mutex locking, and make all access go through them.

Reply:
Should be done!